File: scripts/osm_de_import.py
# OSM-Overpass Deutschland-Import je Bundesland -> public/data/de/{XX}.json
import urllib.request, urllib.parse, json, time, os
OUT = r'C:\Users\Startklar\Documents\Claude\Projects\wasserlage3-src\public\data\de'
os.makedirs(OUT, exist_ok=True)
LAENDER = ['SH','MV','NI','HH','HB','NW','RP','HE','BW','BY','SL','SN','ST','TH','BE','BB']
Q = '''[out:json][timeout:120];area["ISO3166-2"="DE-{L}"]->.a;(
  nwr["leisure"="marina"](area.a);
  nwr["leisure"="slipway"](area.a);
  nwr["waterway"="fuel"](area.a);
  nwr["lock"="yes"](area.a);
  nwr["amenity"="sanitary_dump_station"](area.a);
  nwr["amenity"="boat_rental"](area.a);
  nwr["leisure"="bathing_place"](area.a);
  nwr["mooring"="guest"](area.a);
);out center tags;'''
def kind(t):
    if t.get('leisure')=='marina': return 'hafen'
    if t.get('leisure')=='slipway': return 'slip'
    if t.get('waterway')=='fuel': return 'tank'
    if t.get('lock')=='yes': return 'schleuse'
    if t.get('amenity')=='sanitary_dump_station': return 'entsorgung'
    if t.get('amenity')=='boat_rental': return 'charter'
    if t.get('leisure')=='bathing_place': return 'badestelle'
    if t.get('mooring')=='guest': return 'anleger'
    return None

for L in LAENDER:
    dst = os.path.join(OUT, L + '.json')
    if os.path.exists(dst):
        print(L, 'skip (exists)', flush=True); continue
    body = 'data=' + urllib.parse.quote(Q.replace('{L}', L))
    try:
        req = urllib.request.Request('https://overpass-api.de/api/interpreter', data=body.encode(),
            headers={'Content-Type': 'application/x-www-form-urlencoded', 'User-Agent': 'Wasserlage/3.0 (voigtcarsten4@gmail.com)'})
        j = json.load(urllib.request.urlopen(req, timeout=180))
    except Exception as ex:
        print(L, 'ERROR', ex, flush=True); time.sleep(30); continue
    feats = []
    for el in j.get('elements', []):
        t = el.get('tags', {}); k = kind(t)
        lat = el.get('lat') or (el.get('center') or {}).get('lat')
        lon = el.get('lon') or (el.get('center') or {}).get('lon')
        if not k or not lat or not lon: continue
        nm = t.get('name') or (k=='schleuse' and 'Schleuse') or (k=='slip' and 'Slipanlage') or (k=='badestelle' and 'Badestelle') or (k=='entsorgung' and 'Entsorgungsstation') or (k=='tank' and 'Bootstankstelle') or (k=='anleger' and 'Gastanleger') or None
        if not nm: continue
        tags = [x for x in [t.get('power_supply') and '230V', t.get('drinking_water')=='yes' and 'Trinkwasser',
                t.get('sanitary_dump_station') and 'Entsorgung', t.get('fee')=='no' and 'kostenlos'] if x]
        feats.append(dict(type='Feature', geometry=dict(type='Point', coordinates=[round(lon,5), round(lat,5)]),
            properties=dict(id=f"osm-{el['type']}-{el['id']}", kind=k, name=nm[:80], area=t.get('addr:city') or '',
            region=None, land=L, desc=None, tags=tags, tel=t.get('phone'), web=t.get('website'),
            source='OpenStreetMap', source_detail='Overpass ' + time.strftime('%Y-%m-%d'),
            quality='unverified', coord_quality='exact')))
    json.dump(dict(type='FeatureCollection', features=feats), open(dst, 'w', encoding='utf-8'), ensure_ascii=False)
    print(L, len(feats), flush=True)
    time.sleep(20)
print('ALL DONE', flush=True)

File: workers/osm-de-import.js
/* ═══ Wasserlage 3.0 · OSM-DE-Import-Worker (Cloudflare Cron, Phase C vorbereitet) ═══
 * Importiert je Nacht EIN Bundesland aus Overpass (fair use) in Supabase poi.
 * Deploy: wrangler deploy workers/osm-de-import.js --name wasserlage-osm-import
 * Secrets: wrangler secret put SUPABASE_SERVICE_KEY · SUPABASE_URL · TRIGGER_KEY
 * Cron: "0 2 * * *" (Land rotiert nach Kalendertag)
 */
const LAENDER = [
  { code: 'BE', bbox: '52.33,13.08,52.69,13.77' },
  { code: 'BB', bbox: '51.36,11.27,53.56,14.77' },
  { code: 'MV', bbox: '53.11,10.59,54.69,14.41' },
  { code: 'SH', bbox: '53.36,7.87,55.06,11.31' },
  { code: 'NI', bbox: '51.29,6.65,53.89,11.6' },
];
const QUERY = (bbox) => `[out:json][timeout:90];(
  nwr["leisure"="marina"](${bbox});
  nwr["leisure"="slipway"](${bbox});
  nwr["waterway"="fuel"](${bbox});
  nwr["lock"="yes"](${bbox});
  nwr["amenity"="sanitary_dump_station"](${bbox});
);out center tags;`;
const KIND = (t) =>
  t.leisure === 'marina' ? 'hafen' :
  t.leisure === 'slipway' ? 'slip' :
  t.waterway === 'fuel' ? 'tank' :
  t.lock === 'yes' ? 'schleuse' :
  t.amenity === 'sanitary_dump_station' ? 'entsorgung' : null;

export default {
  async scheduled(_ev, env, ctx) { ctx.waitUntil(run(env)); },
  async fetch(req, env) {
    if (new URL(req.url).searchParams.get('key') !== env.TRIGGER_KEY) return new Response('forbidden', { status: 403 });
    return new Response(JSON.stringify(await run(env)), { headers: { 'content-type': 'application/json' } });
  }
};

async function run(env) {
  const land = LAENDER[new Date().getDate() % LAENDER.length];
  const r = await fetch('https://overpass-api.de/api/interpreter', {
    method: 'POST', body: 'data=' + encodeURIComponent(QUERY(land.bbox)),
    headers: { 'content-type': 'application/x-www-form-urlencoded', 'user-agent': 'Wasserlage/3.0 (voigtcarsten4@gmail.com)' }
  });
  if (!r.ok) return { land: land.code, error: r.status };
  const j = await r.json();
  const rows = (j.elements || []).map(el => {
    const t = el.tags || {}, kind = KIND(t);
    const lat = el.lat ?? el.center?.lat, lon = el.lon ?? el.center?.lon;
    if (!kind || !lat || !lon || !t.name) return null;
    return {
      id: `osm-${el.type}-${el.id}`, kind, name: t.name, lat, lon,
      land: land.code, source: 'OpenStreetMap', source_detail: `Overpass ${new Date().toISOString().slice(0, 10)}`,
      quality: 'unverified', coord_quality: 'exact',
      tel: t.phone || null, web: t.website || null,
      tags: [t.power_supply && '230V', t.drinking_water === 'yes' && 'Trinkwasser'].filter(Boolean)
    };
  }).filter(Boolean);
  /* Upsert in Tranchen à 500 — kuratierte Einträge werden NIE überschrieben (ignore-duplicates) */
  let n = 0;
  for (let i = 0; i < rows.length; i += 500) {
    const res = await fetch(`${env.SUPABASE_URL}/rest/v1/poi?on_conflict=id`, {
      method: 'POST',
      headers: { apikey: env.SUPABASE_SERVICE_KEY, Authorization: `Bearer ${env.SUPABASE_SERVICE_KEY}`,
        'content-type': 'application/json', Prefer: 'resolution=ignore-duplicates,return=minimal' },
      body: JSON.stringify(rows.slice(i, i + 500))
    });
    if (res.ok) n += Math.min(500, rows.length - i);
  }
  return { land: land.code, fetched: rows.length, upserted: n };
}
